Improvements to download as zip

octachrome opened the pull request:
@nightwing As discussed, selecting several files/folders and downloading them results in a file named `file1[+n].zip` where `n` is the total number of files/folders selected.

Using a filename containing `[` and `]` revealed some problems with URI encoding, so I fixed those too. I'm using `encodeURI` to encode the path component of the download URL, and `encodeURIComponent` for the `download` parameter. This seems to give the correct behaviour across a variety of unusual filename characters, including `!$%^&()-_=+[]{};@#~,`.

nightwing commented on plugins/c9.vfs.server/download.js:
```diff
@@ -113,10 +113,8 @@ define(function(require, exports, module) {
-                        path = path.replace(/\\/g, '/');
-                    }
+                    // Single quote the path to escape unusual characters, and manually escape single quotes.
+                    path = "'" + path.replace(/'/, "'\\''") + "'";
```
this breaks downloading as tar.gz on windows

octachrome replied:
@nightwing In which case does it break? It is working fine for me across many test cases, but I have some other shells installed (Git Bash etc.) so it might be that my environment is different to yours.

nightwing replied:
this does not work on linux either

run
```js
mkdir -p 0-ws/x/y/z/
mkdir -p 0-ws/a
echo 1 > 0-ws/x/y/z/z-file
echo 2 > 0-ws/x/y/y-file
echo 3 > 0-ws/x/x-file
echo 4 > 0-ws/root-file
echo 5 > "0-ws/root'quote"
find 0-ws
```
to create file like
```txt
0-ws
0-ws/root'quote
0-ws/x
0-ws/x/y
0-ws/x/y/z
0-ws/x/y/z/z-file
0-ws/x/y/y-file
0-ws/x/x-file
0-ws/root-file
0-ws/a
```

then try to download `0-ws/x/y/z/z-file` and `0-ws/root-file`. that doesn't work for both zip and tar, because node itself quotes the arguments.

nightwing commented on plugins/c9.vfs.client/vfs_client.js:
```diff
@@ -193,12 +193,12 @@ define(function(require, exports, module) {
                 path = path[0];
                 extraPaths = "," + extraPaths.map(function(p) {
-                    return p[0] == path[0] && p != path ? escape(p) : "";
+                    return p[0] == path[0] && p != path ? encodeURI(p) : "";
```
for which filenames escape was giving wrong results?
Asking because without `escape` filenames with `,` will be interpreted as two different filenames also vfsUrl bellow still uses escape.

octachrome replied:
@nightwing If the filename parameter contained a `+`, that was being converted to a space in the downloaded filename, because `escape` leaves `+` alone and `+` encodes a space in URL parameters (but not URL paths). Note that filenames in the path portion of the URL were not affected (i.e., downloading one file containing a `+` works ok, but downloading a folder containing a `+`, or several files of which the first contains a `+`, would give the wrong filename).

I left `vfsUrl` alone because changing it would risk impacting many other VFS calls that I have little knowledge about and cannot test effectively.

octachrome replied:
@nightwing I have reproduced the issue where filenames including a `,` get interpreted as several filenames. I'll think about how to resolve this while still preserving the correct behaviour with filenames containing `+`.

octachrome replied:
@nightwing I think the extra paths should explicitly escape commas in filenames, because they are part of the convention of the download API, e.g.:

                extraPaths = "," + extraPaths.map(function(p) {
                    return p[0] == path[0] && p != path ? encodeURI(p).replace(/,/g, '%2C') : "";
                }).filter(Boolean).join(",");

Using `encodeURI` instead of `escape` ensures that other filename characters such as `+` will be correctly URI-encoded.

nightwing replied:
You are right, touching the escape in vfsUrl is risky.
But with this server calls unescape on result of encodeURI which is not correct.
Try running the following in the browser

```js
[
, original = "/as/f+%2C/аաe"
, encoded = encodeURI(original)
, encodedAndUnescaped = unescape(encoded)
, encodedAndUnescaped == original
, escaped = escape(original)
, escapedAndUnescaped = unescape(escaped)
, original == escapedAndUnescaped
]
```
Since error was only in filename, then

```diff
-                + (filename ? "=" + escape(filename) : "")
+                + (filename ? "=" + encodeURIComponent(filename) : "")
```

should be enough right?

octachrome commented:
@nightwing I think I have fixed all the issues now. I switched back to `escape` because `encodeURIComponent` doesn't escape enough characters, so I am explicitly escaping `+` to avoid it being converted into a space.

I have tested with filenames containing the following characters: ``@~#'`;.,!$%^&()-_=+[]{}``, plus `<>:"\|?*` on Linux, both with single-file and multi-file downloads.

nightwing commented:
merged with a small modification: https://github.com/c9/core/commit/60e4de8ec9ea90e26fdafebb276d1f6165cd546d
Thanks!
